File: Borisova/655.cpp
class Solution {
public:
    int Depth(TreeNode *root) {
        if (!root){
            return 0;
        }
        return 1+max(Depth(root->left),Depth(root->right));
    }
    void dfs(TreeNode *root, vector<vector<string>> &res, int start, int end, int d){
        if (!root) return;
        int M = (start + end) / 2;
        res[d][M] = to_string(root->val);
        d++;
        dfs(root->left, res, start, M, d);
        dfs(root->right, res, M + 1, end, d);
    }
    vector<vector<string>> printTree(TreeNode* root) {
        int d = Depth(root);
        int n = pow(2,d) - 1;
        vector<vector<string>> res(d, vector<string>(n));
        dfs(root, res, 0, n, 0);
        return res;
    }
};

File: Бабат/655.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def printTree(self, root: Optional[TreeNode]) -> List[List[str]]:
        height = self.getHeight(root)
        rows = height
        cols = (2 ** height) - 1
        res = [["" for _ in range(cols)] for _ in range(rows)]
        mid_col = (cols - 1) // 2
        res[0][mid_col] = str(root.val)
        
        self.placeNodes(root.left, res, 1, 0, mid_col - 1, cols)
        self.placeNodes(root.right, res, 1, mid_col + 1, cols, cols)
        
        return res

    def getHeight(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return 1 + max(self.getHeight(root.left), self.getHeight(root.right))
    
    def placeNodes(self, node: Optional[TreeNode], res: List[List[str]], row: int, left: int, right: int, cols: int) -> None:
        if not node:
            return
        mid_col = (left + right) // 2
        res[row][mid_col] = str(node.val)
        self.placeNodes(node.left, res, row + 1, left, mid_col - 1, cols)
        self.placeNodes(node.right, res, row + 1, mid_col + 1, right, cols)
